### ISSUE TITLE
Random jumps between G3 and G4

### ISSUE THREAD

**Issue by pablogila** [thetwom/Tuner#88]
When tuning the 4th string of a Spanish guitar, it indicates G3 (correct) and suddenly it jumps to G4. It does not happen with the rest of the strings.

**Comment by thetwom** [comment]
Hi, thanks for the report. I cannot say how easy is to get around this problem. Sometimes the underlying sound looses most of the characteristica of the lower tone. But it would be good to get a sound sample. The tuner app has the option to record samples and store it to a wave file (please read especially point 4 carefully to understand that the sound BEFORE pressing is stored, not AFTER):
1. Go to preferences and scroll down. You will find the option of recording duration in scientific mode. Turn this up to maximum.
2. Switch on scientific mode.
3. Go back to the tuner, you will now see a record button at the bottom.
4. Pluck you string and wait for the phenomena to happen. After it happened, let it sound for e.g. 1 more second, then press the record button. It will save the 5 seconds BEFORE pressing the button.
5. Give a filename and store it somewhere.
6. Better listen to the stored file before sharing to make sure, it is really something you want to share. 
7. Provide the file here (maybe you have to zip it before ...

If possible, also provide sound samples of the other strings. Maybe also several samples of the 4th string. 

Then I could analyze the sounds and see if improvements are possible. 

Let me know, in case of questions.

**Comment by pablogila** [comment]
Hello! Thanks for your reply. Here are the samples, I hope the naming is self-explanatory. I catched some other notes failing when recorded from an arm-distance. The close samples were recorded with the phone on top of the guitar.
[tuner_issue_#88_samples.zip](https://github.com/user-attachments/files/16603436/tuner_issue_.88_samples.zip)

**Comment by thetwom** [comment]
Thanks a lot. I quickly listened to the samples and think that they will be very helpful to analyse the issue. Can't promise that I can solve it, but I will at least try :-).

**Comment by thetwom** [comment]
Very interesting data. The second harmonic (or overtone) is extremely dominant, leading to the observed behaviour. But the characteristic of the fundamental tone ( the one you expect) is clearly there. 

Here, e.g. the picture for E2, where there is a very dominant peak at ~160Hz (which corresponds to E3). But there are also peaks at multiples of ~80Hz means that E2 is the fundamental. Having dominant frequencies is normally no problem, but in this case it just very dominant :-).
![image](https://github.com/user-attachments/assets/ddd3e5c9-ad66-4023-a0e5-273c2e215106)

In summary, this makes a perfect testcase to better improve the tuner behaviour, so thanks for providing this.

For the time being you can try plucking the string at other positions (especially closer to the center) to trigger less overtones. This of course does not solve the shortcomings of the tuner, but maybe makes it more usable for now :-).

**Comment by pablogila** [comment]
Wow! This is such an interesting graph, thanks for sharing. I did not think the pluck position would matter that much, that is a pretty good point. I will do some tests with different positions, will let you know how it turns out. May I ask how did you process the audio files to plot the frequencies?

**Comment by thetwom** [comment]
> May I ask how did you process the audio files to plot the frequencies?

Do you refer software or the underlying method? About software, these are python-based scripts that I am using to analyse different samples in order to avoid improving the tuner for one instrument on cost of another. So with each new sample of other instruments, I can get a better picture of the performance. Sharing the scripts would not be an issue, but I also guess not as helpful, since they are not made user friendly for end users ;-).

 About the method, the shown graph is the Fourier transform of a small part of your sound sample. The Fourier transform is one of the most basic algorithms in signal analysis and gives a very good picture about the characteristics of a tone. 

The difficulty is  to find a reliable way analyse the resulting Fourier transform :-). The used algorithms seems to work well quite often. But as you see, there is enough room for improvement :-).

**Comment by pablogila** [comment]
That's interesting. I figure that the spectra graph from the science mode is a fourier decomposition. I was able to record some samples by plucking in the middle of the string, and that way the detection worked properly. Plucking at 1/4th the second harmonic dominated just as in the picture you previously provided. Here are the samples, in case you find them useful:
[tuner_plucked_in_the_middle.zip](https://github.com/user-attachments/files/16681851/tuner_plucked_in_the_middle.zip)

It sounds reasonable to avoid improving one instrument on cost of others. However, when not in science mode but in each specific instrument mode, would it make sense to take into account that second harmonics may appear? For example, we know that G3 is the expected harmonic in a 6-string guitar, so if the footprint of a G4 is detected, it could be understood that we are working with G3, etc.

**Comment by thetwom** [comment]
Thanks a lot for the further testing. It is good to know that plucking at the middle changes the picture as expected. 

Yes, exactly, the spectrum in science mode is the Fourier decomposition.  It is used in combination with the second graph (autocorrelation) to find the most probable (not from a mathematical sense of view) and accurate tone. 

> However, when not in science mode but in each specific instrument mode, would it make sense to take into account that second harmonics may appear?

Ideally, this would not be needed. In my opinion consistency is fundamental for user experience, meaning that if you pluck a note while having one instrument set, the same tone is detected with another instrument set (of course this doesn't help, if the tuner can' t decide where to go :-D).  Also it might be very hard to adjust correctly ... . This said, your suggestion is of course still perfectly valid. Why not use extra knowledge if it is there :-). 

For now I have still hopes that by using the new insights from your data I can  improve the tuner behaviour in general. As shown, the data you provided shows a clear trend to the actual tone. At least for one tone I can adjust the some parameters, and it will work very well. Unfortunately, it makes many other samples worse :-). So the "only" thing is to combine the knowledge such it works all the time :D.  I have some ideas to do this, but it needs some time to test.

**Comment by pablogila** [comment]
I am glad that this conversation was fruitful, I found it quite interesting. If you need more samples in a certain setup, I will be happy to record them  :D

**Comment by thetwom** [comment]
So, I tried around a lot. Hopefully, I came up with a significant improvement. At least for the different sound samples, the situation is much better, now but to be honest, my testing uses not exactly the same code base as the tuner (but a very similar one), so there still can be issues. 

Maybe you are willing to test the version: [v7.0.0-rc2](https://github.com/thetwom/Tuner/releases/tag/v7.0.0-rc2)
But please note that this version has a major version jump, so you can not safely go back to a v6....-version without loosing your settings and custom instruments. So, if you have custom instruments, better export them before. About the settings, you would have to remember them :-).

**Comment by Nixola** [comment]
Hi, I just downloaded the app and I'm encountering this issue on v8.0.0; the detected note for the G string jumps up to G4 quite soon and stays there, with a similar thing also happening on the low E string, with sudden (but shorter) octave jumps.  
I'll provide recordings later (can't right now), but the expert mode spectrogram looks similar to the one you posted above.

**Comment by thetwom** [comment]
Hi, thanks for the info. Am I right, that this is *not* a regression from a previous version but an observation you made with this app in general? After introducing v7... there have been no changes in the detection algorithms, so I would not expect behaviour changes lately. 

As you already mentioned, it would be helpful to get sound samples. Also as you say, it is very likely that the spectra look similar as above and that this is the reason why the app struggles getting the correct octave. The underlying algorithms can still deal with such cases, but obviously not always. So lets see whether this can be improved.

**Comment by Nixola** [comment]
Hi, sorry it took me a while, had an unforeseen lack of time.  
I just installed v8.0.0, tried it, and reported the issue; never used it before then.  
For some reason I found it somewhat harder to reproduce than a few days ago, possibly a slightly different background soundscape? At any rate, I'm attaching two samples; one (full) is a full note, from pluck to end, with the issue mostly coming up near the end of the note, while the other (cut) starts a bit later, but the issue was more present throughout the entire note.  
I found the issue much easier to reproduce with the guitar laying horizontally and the phone on top of it, with the microphone near the strings, which is how I recorded the cut sample; with the full sample I was sitting and holding the guitar normally, with the phone laying on my thigh.

[tuner-samples.zip](https://github.com/user-attachments/files/20803454/tuner-samples.zip)

**Comment by thetwom** [comment]
Thanks a lot for the clarifications and the sound samples. I will try to analyse the samples in the next days. Listening to them already gives an impression that these samples are indeed difficult to handle. After ~2-3 seconds (depending on the sample) one can hardly hear the fundamental pitch anymore. As mentioned before you could try to trigger less higher harmonics by plucking more to the center of the string, but this should not be an excuse for the tuner ;-). I will try to understand better what is going on and maybe these samples can help to improve the robustness further.

**Comment by thetwom** [comment]
So, definitely these are sounds which are quite hard to detect correctly, e.g. 

![Image](https://github.com/user-attachments/assets/ad294d68-0e74-4ac6-9d9e-838717601529)

See the huge peak at ~400Hz, which is the second harmonic. The only hint that this is the second and not the first harmonic is the small peak shortly before 600Hz. I think it would be even difficult for a human to detect the right tone, when he starts listening to the sound sample at the second part. 

But this of course also makes the case interesting. At the beginning, the tone can be clearly detected, later it becomes difficult. Meaning, the tuner could try to track the changes of the tone with time and if the tone is very similar to the previous sample, it could suppress an octave jump.. This is something, I largely avoided up to now, since it could of course also lead to unwanted "lock-in" effects, where the tuner would stay at an incorrect note, just because an unfortunate missdetection at the beginning. 

So in summary: Thanks for the samples, outlining weak points of the tuner. Such input really helps to improve the robustness of the tuner. Unfortunately, I do not think that I can provide a short fix for that, so you have to give me some time.  

In case you want to use the app nevertheless for now, you could try to "support" it by plucking the strings in a friendlier way ;-), so more at the centre of the string.

**Comment by thetwom** [comment]
I did some minor improvements now in release [v8.2.0](https://codeberg.org/thetwom/Tuner/releases/tag/v8.2.0) which could help in certain cases, at least for one of your samples (I did not test the other yet).  Let me know whether this shows  improvements. The release should become available on google play and fdroid in the next days.

Note that development shifted now from github.com to codeberg.org. That is why the release link points there. However, you still can keep the discussion here.
